scripts/python/idc_index_data_manager.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import annotations
+
+import logging
+import os
+import re
+import uuid
+from pathlib import Path
+
+import pandas as pd
+from google.cloud import bigquery
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+
+class IDCIndexDataManager:
+    def __init__(self, project_id: str):
+        """
+        Initializes the IDCIndexDataManager.
+
+        Args:
+            project_id (str): The Google Cloud Platform project ID.
+        """
+        self.project_id = project_id
+        self.client = bigquery.Client(project=project_id)
+        logger.debug("IDCIndexDataManager initialized with project ID: %s", project_id)
+
+    def retrieve_latest_idc_release_version(self) -> int:
+        """
+        Retrieves the latest IDC release version from BigQuery.
+
+        Returns:
+            int: The latest IDC release version.
+        """
+        query = """
+        SELECT
+            MAX(idc_version) AS latest_idc_release_version
+        FROM
+            `bigquery-public-data.idc_current.version_metadata`
+        """
+        query_job = self.client.query(query)
+        result = query_job.result()
+        latest_idc_release_version = int(next(result).latest_idc_release_version)
+        logger.debug(
+            "Retrieved latest IDC release version: %d", latest_idc_release_version
+        )
+        return latest_idc_release_version
+
+    def update_sql_queries_folder(
+        self, dir_path: str, latest_idc_release_version: int
+    ) -> None:
+        """
+        Updates SQL queries in the specified folder.
+
+        Args:
+            dir_path (str): The path to the folder containing SQL queries.
+            latest_idc_release_version (int): The latest IDC release version.
+        """
+        for file_name in os.listdir(dir_path):
+            if file_name.endswith(".sql"):
+                file_path = Path(dir_path) / file_name
+                with Path(file_path).open("r") as file:
+                    sql_query = file.read()
+                modified_sql_query = sql_query.replace(
+                    f"idc_current",
+                    f"idc_v{latest_idc_release_version}",
+                )
+                with Path(file_path).open("w") as file:
+                    file.write(modified_sql_query)
+                logger.debug("Updated SQL queries in file: %s", file_path)
```

Comment:
Similarly to the other query, what about adding `idc_index.sql` directly in the this script in a constant called `QUERY_TEMPLATE_GENERATE_IDC_INDEX`, and since parameterized query cannot[^1] be used for table name, you would have a place holder `@IDC_TABLENAME@`.

Then you would simply do the following from an updated function `generate_index_data_files`:

```python
def generate_index_data_files(self, idc_version: int, output_basename: str, to_compressed_csv: bool=True, to_parquet: bool=False):

    query = QUERY_TEMPLATE_GENERATE_IDC_INDEX.replace("@IDC_TABLENAME@", f"idc_v{idc_version})

    index_df = self.client.query(query)
    
    if to_compressed_csv:
        csv_file_name = f"{output_basename}.csv.zip"
        index_df.to_csv(csv_file_name, compression={"method": "zip"}, escapechar="\\")
        logger.debug("Created CSV zip file: %s", csv_file_name)
    
    if to_parquet:
        parquet_file_name = f"{output_basename}.parquet"
        index_df.to_parquet(parquet_file_name)
        logger.debug("Created Parquet file: %s", parquet_file_name)
```


[^1]: https://cloud.google.com/bigquery/docs/parameterized-queries

Reply:
In near feature, we will have more than one query. i.e a query in addition to generic idc_index.sql, an index specific to SlideMicroscopy (SM) modality for Pathology. So I think we may want to have queries in its own .sql file as we have now. 
@fedorov what's your take on this?

Reply:
To address you could have another template `QUERY_TEMPLATE_GENERATE_IDC_INDEX_SLIDE_MICROSCOPY`, I think having the query template along side the code remove indirection and simplify the overall infrastructure.

In the future, if these "index-data" archive have different release pace, you could even consider having wheels like `idc-index-data-slide-miscrosopy` or `idc-index-data[slide-miscrosopy]`

Also, you could even have a CLI within this package for generating the sql query. For example, `python -m idc-index-data --output-sql-query --version 17`

Reply:
All of them are great points! I guess the only downside is if someone is interested in which queries were used to generate the indices, they will have to dig the python script.

Reply:
> which queries were used to generate the indices, they will have to dig 

Since the current sql file are not the one effectively used to generate the index, you would have to explain the user how to interpret the query and update it so that it corresponds to the one that was used.

To properly address this, I then suggest:
- CLI like `python -m idc-index-data --output-sql-query`: This would output the query effectively used to generate the current index
- publish documentation at `idc-index-data.readthedocs.io` with the corresponding queries and generate information about the infrastructure and schema related to bigquery tables

Reply:
Thank you! Let's see what Andrey also thinks. 
In the mean time, I'll work on the changes you recommended tomorrow and hand the PR over to you to address any remaining enhancements that I do not fully know yet.

Reply:
We would not be replacing the table name, but the dataset name, and the dataset name in the non-versioned query is always `idc_current`, so I think we can keep this simple.

Reply:
In fact, we can simplify this even further. There is no need to replace `idc_current` with the versioned dataset for the purposes of query execution. The result will be identical. The only reason we need to map `idc_current` to the version is to keep track of the IDC version for the provenance of the query generation at a later time, when `idc_current` is different. We can even skip saving the versioned queries anywhere - the version of IDC will be captured in the version of the PyPI package. How about we make it this simple for now, and refine if needed in the future?

Manually triggered workflow:

1. run each of the queries "as is"
2. save CSV and Parquet results for each query as GHA artifacts

Workflow triggered on release tag creation:

1. run each of the queries "as is"
2. publish PyPI package with the attached CSV and/or Parquet results for each query

This should make the code a lot simpler, reducing maintenance, and I do not think we lose much in terms of functionality.

Reply:
Sure, the latest iteration does all the things you mentioned.